### PR TITLE
Create abstract Verifier and Signer, remove key_id hack from App Engine and IAM signers

### LINKED ISSUE #112: Consider allowing Signer.key_id to be None
#108 and #109 added a "hack" to allow the key id to be determined when signing with IAM and App Engine. The reasoning for this is that the key id is needed to populate the `kid` claim in the JWT header.

It turns out, the `kid` field is optional according to the JWT spec. If it's absent, it's undefined what to do with it, but it seems that most clients will try all available certs (which is what we do as well).

@dhermes @lukesneeringer what do you think about allowing key_id to be None for `iam.Signer` and `app_engine.Signer`?

### PR DESCRIPTION
**Pull request by jonparrott**
Resolves #112

### CODE REVIEW

File: google/auth/app_engine.py [hunk 1]
```diff
@@ -26,49 +26,33 @@
...
     running in the App Engine standard environment.
-
-    .. warning::
-        The App Identity service signs bytes using Google-managed keys.
-        Because of this it's possible that the key used to sign bytes will
-        change. In some cases this change can occur between successive calls
-        to :attr:`key_id` and :meth:`sign`. This could result in a signature
-        that was signed with a different key than the one indicated by
-        :attr:`key_id`. It's recommended that if you use this in your code
-        that you account for this behavior by building in retry logic.
```
Comment (dhermes) [inline google/auth/app_engine.py]: Why did this die?

File: google/auth/crypt.py [hunk 2]
```diff
@@ -193,6 +201,33 @@ def sign(self, message):
         Returns:
             bytes: The signature of the message.
         """
+        # pylint: disable=missing-raises-doc,redundant-returns-doc
+        # (pylint doesn't recognize that this is abstract)
+        raise NotImplementedError('Sign must be implemented')
+
+
+class RSASigner(Signer):
+    """Signs messages with a RSA private key.
```
Comment (dhermes) [inline google/auth/crypt.py]: You said "an RSA" above. Which is it: "a" or "an"?

File: google/auth/crypt.py [hunk 3]
```diff
@@ -193,6 +201,33 @@ def sign(self, message):
...
+        private_key (rsa.key.PrivateKey): The private key to sign with.
+        key_id (str): Optional key ID used to identify this private key. This
+            can be useful to associate the private key with its associated
+            public key or certificate.
+    """
+
+    def __init__(self, private_key, key_id=None):
+        self._key = private_key
+        self._key_id = key_id
+        """"""
```
Comment (dhermes) [inline google/auth/crypt.py]: Sextuple quotes! WAT?

File: google/auth/app_engine.py (hunk 1 above)
Comment (jonparrott) [inline google/auth/app_engine.py]: Because we no longer even try to get the key ID.

File: google/auth/crypt.py (hunk 2 above)
Comment (jonparrott) [inline google/auth/crypt.py]: `an`. Fixed.

File: google/auth/crypt.py (hunk 3 above)
Comment (jonparrott) [inline google/auth/crypt.py]: Removed, remnants of a docstring.

**Comment by dhermes** [comment]
LGTM?

**Comment by jonparrott** [comment]
> LGTM?

6-anchorman-quotes-burgundy

**Pull request merged**